### DISCUSSION TITLE
iPXE on IIS

### DISCUSSION THREAD

**Discussion by christoph-wijns** [ipxe/ipxe#217] [General]
Hello NOOB here ;)

Is there a good tutorial to configure iPXE on IIS ?

Many thanks

**Comment by NiKiZe** [comment]
What is it that you need help with?

There should be no or little difference to any other http server. With maybe the issue that IIS does not allow files without extension or unknown extensions by default.

You might however want to get the parts prior to http running first by testing agains http://boot.ipxe.org/demo

**Reply by christoph-wijns** [reply]
Thanks for answering ...

What I see/know is the content of the folder (virtual directory) ?

Many thanks

**Comment by NiKiZe** [comment]
To get iPXE running you don't need any http server at all, after that it all depends on what you want to do. There is no default or standard iPXE share

**Reply by christoph-wijns** [reply]
> To get iPXE running you don't need any http server at all, after that it all depends on what you want to do. There is no default or standard iPXE share

What I want to see is like a printscreen of folder structure

I like to achieve a menu 

It is for a starter so please ;)

Thank you

**Reply by NiKiZe** [reply]
Again there is no standard.
You only need a ipxe script, or not even that
but as an example, create a boot.ipxe on your webserver, see https://ipxe.org/scripting for what to do in that file.
For details on menu see https://ipxe.org/cmd/menu

**Reply by christoph-wijns** [reply]
Well I'm a little bit further - I've implemented PDS MDT https://github.com/FriendsOfMDT/PSD (Johan Arwidmark)

MDT webserver (image)

![ipxe_forum_01](https://user-images.githubusercontent.com/75322374/106157026-8a5e2300-6182-11eb-82e4-cdc866667a6e.PNG)

iPXE boot.php

 #!ipxe
  
  kernel wimboot
  initrd http://mdt01.chw.labo/psdproduction/Boot/x64/Boot/BCD
  initrd http://mdt01.chw.labo/psdproduction/Boot/x64/Boot/boot.sdi
  initrd http://mdt01.chw.labo/psdproduction/Boot/LitetouchPE.wim
  boot

PXE01 server (image)

![ipxe_forum_02](https://user-images.githubusercontent.com/75322374/106158004-90083880-6183-11eb-957b-e31175312753.PNG)

DHCP Option 67

![ipxe_forum_03](https://user-images.githubusercontent.com/75322374/106158386-f68d5680-6183-11eb-96f4-2a41df619515.PNG)

Not able to get it working ;)

Many thanks for helping

**Reply by mcb30** [reply]
> Not able to get it working ;)

@chw1979 You might get more useful replies if you give some (any) information about the way in which it doesn't work for you.  For example: what do you see on-screen when you try to boot?

There are many guides available on how to submit a useful problem report, such as https://www.chiark.greenend.org.uk/~sgtatham/bugs.html

**Reply by Hammarskjold** [reply]
You cannot use an HTTP path for the option 67 in a MS DHCP server, it needs to be a filename local to the TFTP root on the TFTP server (specified by option 66).

**Reply by christoph-wijns** [reply]
**MCB30** you have totally reason ;)

So far (PSD MDT Johan Arwidmark) implemented

DHCP Microsoft (option 67 - http://mdt01.chw.labo/psdproduction/Boot/ipxe.ipxe)

#!ipxe
set boot-url http://mdt01.chw.labo/psdproduction
kernel ${boot-url}/Boot/ipxe/wimboot
initrd ${boot-url}/Boot/x86/Boot/BCD BCD
initrd ${boot-url}/Boot/x86/Boot/boot.sdi boot.sdi
initrd ${boot-url}/Boot/LiteTouchPE_x86.wim LiteTouchPE_x86.wim
boot

Due to covid I'm not able to test it on a physically machine but with XCP-NG - create a VM with Boot firmware BIOS (legacy)

![ipxe_forum_04](https://user-images.githubusercontent.com/75322374/106477436-72e7a880-64a8-11eb-801f-11ebe5c1f8eb.PNG)

So I think everything goes well but al of a sudden the VM shuts down ...

Maybe due to XCP-NG virtualization ?

Many thanks

**Reply by mcb30** [reply]
> So I think everything goes well but al of a sudden the VM shuts down ...

OK, and what's the **very last** thing you see displayed on the screen just before it shuts down?

For example: is it always at 22% of the way through downloading `LiteTouchPE_x86.wim`, as in your screenshot?  Does it reboot in the middle of the download, or does it reach the end of the download and then reboot?

**Reply by christoph-wijns** [reply]
It reaches the end - 100% but then the VM shuts down - I'm not seeing anything else ...

Tried troubleshooting features but also no info

**Reply by mcb30** [reply]
> It reaches the end - 100% but then the VM shuts down - I'm not seeing anything else ...

That is useful information, thank you.  Please update your iPXE script to:

```
#!ipxe
set boot-url http://mdt01.chw.labo/psdproduction
kernel ${boot-url}/Boot/ipxe/wimboot
initrd ${boot-url}/Boot/x86/Boot/BCD BCD
initrd ${boot-url}/Boot/x86/Boot/boot.sdi boot.sdi
initrd ${boot-url}/Boot/LiteTouchPE_x86.wim LiteTouchPE_x86.wim

# Extra commands here
imgargs wimboot pause
imgstat
prompt Press a key to start wimboot
boot
```

That will introduce two pauses into the boot flow: one immediately before iPXE hands off control to wimboot and one immediately before wimboot hands off control to Windows.  Please report what you see at each of those points.

**Reply by christoph-wijns** [reply]
Implemented troubleshooting ...

![ipxe_forum_05](https://user-images.githubusercontent.com/75322374/106515893-988ba680-64d6-11eb-83b7-f22c7c0f5229.PNG)

When I press thé any key VM shuts down

I'm googling ;)

**Reply by christoph-wijns** [reply]
My guess is that the issue is related to XCP-NG virtualization - for many days I read blogs and blogs and always I find the 'same' boot.ipxe config ...

Trying tomorrow on Hyper-V

**Reply by christoph-wijns** [reply]
Hello Again ;)

**Compute**
Hyper-V
Microsoft DHCP
Test-Machine - Legacy & UEFI

I want to boot a **boot.wim** file (extracted from Windows 10 ISO)

**_Actions taken_**

**MIME Type (web.config)**
```
<?xml version="1.0" encoding="UTF-8"?>
<configuration>
    <system.webServer>
        <directoryBrowse enabled="true" />
        <staticContent>
            <mimeMap fileExtension="." mimeType="application/octet-stream" />
        </staticContent>
    </system.webServer>
</configuration>
```

IIS content
![ipxe_forum_01](https://user-images.githubusercontent.com/75322374/106904621-b38b3000-66fb-11eb-9a4f-4f8f75f1ec37.png)


Generate a **undionly.kpxe & snponly.efi** with https://ipxe.pcserviceburgas.com/ (very slow and often unresponsive)
Content of file
```
#!ipxe

set base-url http://ipxe01.chw.labo/ipxe
kernel ${base-url}/wimboot
initrd ${base-url}/BCD         	BCD
initrd ${base-url}/boot.sdi    	boot.sdi
initrd ${base-url}/boot.wim	boot.wim
```

Ps: are there web-alternatives to create those files ?
Rom-O-Matic throws error
![ipxe_forum_07](https://user-images.githubusercontent.com/75322374/106912706-8e022480-6703-11eb-9c8c-c80b91428424.png)


Microsoft DHCP Options
https://community.theforeman.org/t/netbooting-uefi-and-legacy-bios-clients-using-windows-dhcp/19816

```
Add-DhcpServerv4Class -Name "PXEClient:Arch:00007" -Type Vendor -Description "PXEClient (UEFI x64)" -Data "PXEClient:Arch:00007"
Add-DhcpServerv4Class -Name "PXEClient:Arch:00006" -Type Vendor -Description "PXEClient (UEFI x86)" -Data "PXEClient:Arch:00006"
Add-DhcpServerv4Class -Name "PXEClient:Arch:00000" -Type Vendor -Description "PXEClient (Legacy BIOS)" -Data "PXEClient:Arch:00000"
Add-DhcpServerv4Class -Name "iPXE" -Type User -Description "iPXE clients" -Data "iPXE"
```

```
# iPXE client chanloads into bootstrap
Add-DhcpServerv4Policy -ProcessingOrder 1 -Name "chainload-ipxe" -Condition OR -UserClass EQ,iPXE
Set-DhcpServerv4OptionValue -PolicyName "chainload-ipxe" -OptionID 67 -Value "http://ipxe01.chw.labo/ipxe/boot.ipxe"

# UEFI clients load iPXE
Add-DhcpServerv4Policy -ProcessingOrder 2 -Name "pxeclient-uefi64" -Condition OR -VendorClass EQ,PXEClient:Arch:00007*
Set-DhcpServerv4OptionValue -PolicyName "pxeclient-uefi64" -OptionID 67 -Value "http://ipxe01.chw.labo/ipxe/snponly.efi"

# Legacy clients just load usual pxelinix.0
Add-DhcpServerv4Policy -ProcessingOrder 3 "pxeclient-bios" -Condition OR -VendorClass EQ,PXEClient:Arch:00000*
Set-DhcpServerv4OptionValue -PolicyName "pxeclient-bios" -OptionID 67 -Value "http://ipxe01.chw.labo/ipxe/undionly.kpxe"
```

I know there are other architectures as well but for lab & testing let me start with Legacy ;)

![ipxe_forum_03](https://user-images.githubusercontent.com/75322374/106905656-c8b48e80-66fc-11eb-8cb1-e1f73e0cfc9f.png)

![ipxe_forum_04](https://user-images.githubusercontent.com/75322374/106905729-d79b4100-66fc-11eb-851a-2c35b2e9b7d6.png)


Content of **http://ipxe01.chw.labo/ipxe/boot.ipxe** (same as undionly.kpxe & snponly.efi) - maybe overkill
```
#!ipxe

set base-url http://ipxe01.chw.labo/ipxe
kernel ${base-url}/wimboot
initrd ${base-url}/BCD         	BCD
initrd ${base-url}/boot.sdi    	boot.sdi
initrd ${base-url}/boot.wim	boot.wim
```

Result - neither Legacy of UEFI boots on iPXE

Legacy
![ipxe_forum_06](https://user-images.githubusercontent.com/75322374/106906197-395bab00-66fd-11eb-86da-a7e67562d21b.png)

Wireshark
![ipxe_forum_05](https://user-images.githubusercontent.com/75322374/106906210-3c569b80-66fd-11eb-98a7-bebdb55eb5b1.png)

IIS logs shows nothing - my guess - URL option 67 is not properly translated

I'm almost there **I HOPE** but something goes terribly wrong ;)

Anyone advice ...

Many thanks

**Reply by mcb30** [reply]
The problem is that you are providing an HTTP URL for the `pxeclient-bios` and `pxeclient-uefi64` policies:

> ```
> # UEFI clients load iPXE
> Add-DhcpServerv4Policy -ProcessingOrder 2 -Name "pxeclient-uefi64" -Condition OR -VendorClass EQ,PXEClient:Arch:00007*
> Set-DhcpServerv4OptionValue -PolicyName "pxeclient-uefi64" -OptionID 67 -Value "http://ipxe01.chw.labo/ipxe/snponly.efi"
> 
> # Legacy clients just load usual pxelinix.0
> Add-DhcpServerv4Policy -ProcessingOrder 3 "pxeclient-bios" -Condition OR -VendorClass EQ,PXEClient:Arch:00000*
> Set-DhcpServerv4OptionValue -PolicyName "pxeclient-bios" -OptionID 67 -Value "http://ipxe01.chw.labo/ipxe/undionly.kpxe"
> ```

The built-in BIOS and UEFI PXE clients in Hyper-V cannot use HTTP: they are limited to using only TFTP.  You will need to provide a plain filename instead, e.g.:

```
Set-DhcpServerv4OptionValue -PolicyName "pxeclient-uefi64" -OptionID 67 -Value "snponly.efi"
Set-DhcpServerv4OptionValue -PolicyName "pxeclient-bios" -OptionID 67 -Value "undionly.kpxe"
```

and then use a TFTP server to serve up the `snponly.efi` and `undionly.kpxe` files.

Once your Hyper-V VM has loaded `snponly.efi` or `undionly.kpxe` via TFTP, you will be running iPXE and so the next part:

> ```
> # iPXE client chanloads into bootstrap
> Add-DhcpServerv4Policy -ProcessingOrder 1 -Name "chainload-ipxe" -Condition OR -UserClass EQ,iPXE
> Set-DhcpServerv4OptionValue -PolicyName "chainload-ipxe" -OptionID 67 -Value "http://ipxe01.chw.labo/ipxe/boot.ipxe"
> ```

will work as expected, since iPXE **does** know how to use HTTP.

(In your original test setup using XCP-ng you did not have this particular problem, since XCP-ng uses iPXE to provide its built in PXE client, and is therefore capable of immediately using HTTP.)

**Reply by christoph-wijns** [reply]
Thanks for info 

So I need to install tftpd32 (example) on iPXE server

Issue only on Hyper-V - not with physical machine ?

Thanks thanks thanks a lot for helping 

Been struggling a lot - iPXE was on my bucketlist for 'years' and it is not easy - like most of my bucketlist but such a nice feature !!!

**Reply by mcb30** [reply]
> So I need to install tftpd32 (example) on iPXE server

Yes, that should work as an option.

> Issue only on Hyper-V - not with physical machine ?

No, you'll find that most physical machines can also use only TFTP from their built-in PXE ROMs.  A few machines will have iPXE ROMs that can perform HTTP, but these are rare.

> Been struggling a lot - iPXE was on my bucketlist for 'years' and it is not easy - like most of my bucketlist but such a nice feature !!!

Glad to help!

**Comment by christoph-wijns** [comment]
I know - i have experience with pxelinux ... but in our topology I want to try iPXE HTTP(s) but do not find a good installation/setup blog

Many thanks for answering

**Reply by Hammarskjold** [reply]
So then you first get to iPXE (via TFTP regular PXE boot by chainloading iPXE) and then using iPXE to download the files (via the PHP?) via the web service. You are on the right track but missing some key fundamentals.

**Comment by christoph-wijns** [comment]
Back again 😏

I have a working environment for iPXE http but I like to go a step further and use https 

Implemented ca & cross-ca from https://ca.ipxe.org/ and mmc shows that the chain is OK but how do I create the IIS web certificate ?

@this moment I´ve only got a MS environment - in my labo I also have a PKI infrastructure ... for what I understand is that iPXE works out of the box with https://ca.ipxe.org/ so let´s start by understanding that technology/sequence 

Summary - based on cert chain from https://ca.ipxe.org/ how do I create IIS web certificate in a MS environment ?

Many thanks 

Christoph

**Reply by NiKiZe** [reply]
There are many resources available for how to get and install SSL certificates, little or nothing that is specific to iPXE, letsencrypt is also a possible alternative.
